### PR TITLE
DA-376: CBLite CRUD Scopes Collections Documents

### PR DESCRIPTION
**Pull request by teetangh**
Description:
This Pull Request (PR) focuses on expanding the functionality of the Couchbase Lite IntelliJ IDEA plugin to empower users with the ability to create collections and manage scopes and collections efficiently. The primary goal is to provide a seamless experience within the IntelliJ IDEA environment for users working with Couchbase Lite.

Key Changes:

**Collection Creation**: With this update, users will be able to create collections directly through the plugin interface. This feature simplifies the development process and streamlines the management of Couchbase Lite collections.

**Scope and Collection Deletion**: The PR also introduces the capability to delete both scopes and collections from the plugin. Users can now easily remove unnecessary scopes and collections, ensuring a clean and organized database setup.

### CODE REVIEW

**Comment by deniswsrosa** [comment]
Add a description to the PR

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLDataLoader.java [hunk 1]
```diff
@@ -59,6 +59,29 @@ public static SavedCBLDatabase saveNewDatabase(String id, String name, String pa
         return newdDb;
     }
 
+    public static void listScopes(DefaultMutableTreeNode parentNode) throws CouchbaseLiteException {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLDataLoader.java:62]: Never **ever** commit code that is not being used.

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 2]
```diff
@@ -89,58 +100,114 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+                        Log.error(ex);
+                        SwingUtilities
+                                .invokeLater(() -> Messages.showErrorDialog("Could not disconnect from the database.",
+                                        "Couchbase Lite Connection Error"));
+                    }
+                }
+            };
+            actionGroup.add(menuItem);
+
+            AnAction createCollection = new AnAction("Create Scoped Collection") {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:125]: Avoid creating new terms, "Scoped Collection" is not something that mentioned anywhere in our documentation. Rename to "Create Collection"

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 3]
```diff
@@ -89,58 +100,114 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+                    try {
+                        CBLCreateScopedCollectionDialog dialog = new CBLCreateScopedCollectionDialog(project, tree);
+                        dialog.show();
+                        CBLDataLoader.loadScopesAndCollections(clickedNode);
+                        ((DefaultTreeModel) tree.getModel()).nodeStructureChanged(clickedNode);
+                    } catch (Exception ex) {
+                        Log.error(ex);
+                        SwingUtilities
+                                .invokeLater(() -> Messages.showErrorDialog("Could not create the scoped collection.",
+                                        "Couchbase Lite Plugin Error"));
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:137]: Rename the Scoped Collection to "Collection"

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 4]
```diff
@@ -89,58 +100,114 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+                    }
+                }
+            };
+            actionGroup.add(menuItem);
+
+            AnAction createCollection = new AnAction("Create Scoped Collection") {
+                @Override
+                public void actionPerformed(@NotNull AnActionEvent e) {
+                    try {
+                        CBLCreateScopedCollectionDialog dialog = new CBLCreateScopedCollectionDialog(project, tree);
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:129]: "ScopedCollection" should not be a name

File: src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateScopedCollectionDialog.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+public class CBLCreateScopedCollectionDialog extends DialogWrapper {
+    private JPanel panel;
+    private JBTextField scopeField;
+    private JBTextField collectionField;
+    private JLabel errorLabel;
+    private Project project;
+    private Tree tree;
+
+    public CBLCreateScopedCollectionDialog(Project project, Tree tree) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateScopedCollectionDialog.java:30]: "CBLCreateCollectionDialog"

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 6]
```diff
@@ -89,58 +100,114 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
-                        } catch (Exception ex) {
-                            ex.printStackTrace();
-                            Log.debug("Could not close the file", ex);
-                        }
-                    }
-
-                    DefaultMutableTreeNode parentNode = (DefaultMutableTreeNode) clickedNode.getParent();
-                    if (parentNode != null) {
-                        ((DefaultTreeModel) tree.getModel()).removeNodeFromParent(clickedNode);
+                    for (int i = 0; i < clickedNode.getChildCount(); i++) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:183]: This whole code is blocking the UI. You need to run it in a background task

Comment (deniswsrosa) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:183]: Did You read Aaron's spec? Because it says: **NOTE:  Deleting documents that have blob attachments on them will require you to run the database maintenance API with the MaintenanceType of COMPACT.**

Once you finish deleting everything, you need to run this maintenance.  ActiveCBLDatabase.getInstance().getDatabase().performMaintenance(MaintenanceType.COMPACT);

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 7]
```diff
@@ -149,62 +216,108 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+            public void actionPerformed(@NotNull AnActionEvent e) {
+                int result = Messages.showYesNoDialog("<html>Are you sure you want to delete the collection <strong>"
+                        + userObject.getText() + "</strong>?</html>", "Delete Collection", Messages.getQuestionIcon());
+                if (result != Messages.YES) {
+                    return;
+                }
+
+                try {
+
+                    if (ActiveCBLDatabase.getInstance() != null
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:237]: Can you give an example of when ActiveCBLDatabase.getInstance() is null? Why did you add this check?

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 8]
```diff
@@ -149,62 +216,108 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+                            && ActiveCBLDatabase.getInstance().getDatabase() != null) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:238]: How can the user even get to this option if he is not connected to the database?

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 9]
```diff
@@ -149,62 +216,108 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+                        ActiveCBLDatabase.getInstance().getDatabase().deleteCollection(userObject.getText(),
+                                userObject.getScope());
+                        DefaultMutableTreeNode parentNode = (DefaultMutableTreeNode) clickedNode.getParent();
+                        if (parentNode != null && tree.getModel() instanceof DefaultTreeModel) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:242]: I won't even review this code because a lot of checks here are looking really awkward. Feels like a code that was autogenerated. Refactor your code first to something that looks understandable.

**Review by deniswsrosa — CHANGES_REQUESTED** [review]
See comments. Refactor is much needed. Many lines you have changed just the formatting.

File: .idea/gradle.xml [hunk 10]
```diff
@@ -5,7 +5,7 @@
     <option name="linkedExternalProjectsSettings">
       <GradleProjectSettings>
         <option name="externalProjectPath" value="$PROJECT_DIR$" />
-        <option name="gradleJvm" value="jbrsdk_jcef-17" />
+        <option name="gradleJvm" value="jbrsdk_jcef-21" />
```
[the remark this answers is not in the document]
Comment (teetangh) [reply .idea/gradle.xml:8]: Okay am removing it. But CBLite option is not appearing on jcef-17
image

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLDataLoader.java (hunk 1 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLDataLoader.java:62]: okay. removing it.

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java (hunk 8 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java]: removed it

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java (hunk 7 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java]: removed it

**Comment by deniswsrosa** [comment]
- You need to break a line if there is more than one error. Check other validations, we are doing that there already.

Screenshot 2023-11-09 at 17 45 17

**Comment by deniswsrosa** [comment]
- Looks like you are appending errors somewhere in your code.

Screenshot 2023-11-09 at 17 46 57

**Comment by deniswsrosa** [comment]
- Show the name of the blob instead

Screenshot 2023-11-09 at 17 54 19

**Comment by deniswsrosa** [comment]
Screenshot 2023-11-09 at 17 58 46
You are not handling well files with spaces in their names I think

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 11]
```diff
@@ -189,62 +267,286 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
+        AnAction attachBlob = new AnAction("Attach Blob") {
+            @Override
+            public void actionPerformed(@NotNull AnActionEvent e) {
+                try {
+                    Scope scope = ActiveCBLDatabase.getInstance().getDatabase().getScope(userObject.getScope());
+                    Collection collection = Objects.requireNonNull(scope).getCollection(userObject.getCollection());
+                    assert collection != null;
+                    Document document = collection.getDocument(userObject.getId());
+                    CBLAttachBlobDialog dialog = new CBLAttachBlobDialog(project, collection, document);
+                    if (dialog.showAndGet()) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:388]: Historically, we let the dialog refresh the tree. No need to refactor the code, but would be nice if that is done in the dialog next time.

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java [hunk 12]
```diff
@@ -189,62 +267,286 @@ public void actionPerformed(@NotNull AnActionEvent e) {
...
             public void actionPerformed(@NotNull AnActionEvent e) {
                 try {
-                    CBLTreeHandler.deleteConnection(clickedNode, userObject, tree);
-                } catch (CouchbaseLiteException ex) {
-                    Log.error(ex);
-                    SwingUtilities.invokeLater(() -> Messages.showErrorDialog("Could not delete the connection.", "Couchbase Lite Plugin Error"));
+                    // Assuming blob is a file on disk
+                    File blobFile = new File(userObject.getDigest());
+
+                    if (!Desktop.isDesktopSupported()) {
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java:477]: this strategy here needs to change. We need to open it with the editor if possible or not show this option at all according to the file type.

File: src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateCollectionDialog.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        }
+
+        if (collectionValidationError != null || scopeValidationError != null) {
+            return;
+        }
+
+        try {
+            ActiveCBLDatabase.getInstance().getDatabase().createCollection(collectionName, scopeName);
+        } catch (Exception e) {
+            errorLabel.setText("Error creating collection: " + e.getMessage());
```
Comment (deniswsrosa) [inline src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateCollectionDialog.java:122]: Log the error

**Review by deniswsrosa — CHANGES_REQUESTED** [review]
Almost there. There are a few additional things to fix.

**Comment by teetangh** [comment]
> * You need to break a line if there is more than one error. Check other validations, we are doing that there already.
> 
> Screenshot 2023-11-09 at 17 45 17

image

**Comment by teetangh** [comment]
> * Looks like you are appending errors somewhere in your code.
> 
> Screenshot 2023-11-09 at 17 46 57

I guess this one occurred on an older commit.

**Comment by teetangh** [comment]
> * Show the name of the blob instead
> 
> Screenshot 2023-11-09 at 17 54 19

I have fixed this one now

image

File: src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateCollectionDialog.java (hunk 13 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateCollectionDialog.java:122]: Added Logging.

Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/dialog/CBLCreateCollectionDialog.java:122]: image

**Comment by teetangh** [comment]
> Screenshot 2023-11-09 at 17 58 46 You are not validating if the user has specified if the user has specified the blob name

Have fixed this now. Also, fixed the weird behviour of stacking multiple "browse file dialog" has also been resolved 

image

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java (hunk 9 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java]: Have removed all the extra null checks

File: src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java (hunk 6 above)
Comment (teetangh) [reply src/main/java/com/couchbase/intellij/tree/cblite/CBLTreeRightClickListener.java]: Performing maintenance after every scope/collection/document/blob deletion.

**Review by deniswsrosa — APPROVED** [review]
